3rdparty/tools/rust/BUILD

```diff
@@ -0,0 +1,58 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+file(
+    name="rustup-init",
+    source=per_platform(
+        linux_arm64=http_source(
+            url="https://static.rust-lang.org/rustup/dist/aarch64-unknown-linux-gnu/rustup-init",
+            len=14131368,
+            sha256="673e336c81c65e6b16dcdede33f4cc9ed0f08bde1dbe7a935f113605292dc800",
+        ),
+        linux_x86_64=http_source(
+            url="https://static.rust-lang.org/rustup/dist/x86_64-unknown-linux-gnu/rustup-init",
+            len=14293176,
+            sha256="0b2f6c8f85a3d02fde2efc0ced4657869d73fccfce59defb4e8d29233116e6db",
+        ),
+        macos_arm64=http_source(
+            url="https://static.rust-lang.org/rustup/dist/aarch64-apple-darwin/rustup-init",
+            len=8000713,
+            sha256="ed299a8fe762dc28161a99a03cf62836977524ad557ad70e13882d2f375d3983",
+        ),
+        macos_x86_64=http_source(
+            url="https://static.rust-lang.org/rustup/dist/x86_64-apple-darwin/rustup-init",
+            len=8670640,
+            sha256="f6d1a9fac1a0d0802d87c254f02369a79973bc8c55aa0016d34af4fcdbd67822",
+        ),
+    ),
+)
+
+shell_command(
+    name="cargo-shim",
+    command="""
+        chmod +x rustup-init
+        ./rustup-init --no-modify-path --no-update-default-toolchain --default-toolchain none -y
+        mv shims/bin/cargo cargo
+    """,
+    extra_env_vars=["RUSTUP_HOME=dummy", "CARGO_HOME=shims"],
+    tools=["chmod", "mv"],
+    execution_dependencies=[":rustup-init"],
+    output_files=["cargo"],
+)
+
+shell_command(
+    name="cargo",
+    command="""
+        mv ../../../src/rust/engine/rust-toolchain rust-toolchain
+        ./cargo version
```

Comment:
Ah, also one run of this exceeded the default 30s timeout for me, on my 100Mbit/s connection. (A later run worked in 24s, so 🤷‍♂️ )

I estimated the max download size as ~414MiB based on the maximum sizes for each component listed as being installed, for 1.72: `cargo` (18.1), `clippy` (9.97), `rust-docs` (70.25), `rust-std` (62.12), `rustc` (259.21(!)), `rustfmt` (5.17).

If someone has a 5Mbit/s connection, this takes ~12 minutes to download, hence maybe this could have a bit of slack with a timeout like 1200s (20 minutes)?

Reply:
I wish we could avoid downloading the ones we don't need for this operation. I remember now that was a motivation for the earlier art.

Until then, yeah let's just bump it.

Reply:
Yeah, that'd be cool. That said, "fortunately" cargo, rust-std and rustc alone seem to be ~80% of the bytes, so optimising would be saving 20% at most, and thus still need a timeout > 30s for slower connections.